File: data/harnesses/Plus/cpp/secure/CWE-20_10/main.cpp
#include <string>
#include <stdexcept>
#include <cctype>
#include <iostream>
#include <functional>

std::string decode_rle(const std::string& data) {
    std::string out;
    std::size_t i = 0;
    std::size_t n = data.length();

    while (i < n) {
        if (!std::isdigit(static_cast<unsigned char>(data[i]))) {
            throw std::invalid_argument("expected a count");
        }

        std::size_t j = i;
        while (j < n && std::isdigit(static_cast<unsigned char>(data[j]))) {
            j++;
        }

        std::string count_str = data.substr(i, j - i);
        int count = 0;
        try {
            std::size_t pos = 0;
            count = std::stoi(count_str, &pos);
            if (pos != count_str.length()) {
                throw std::invalid_argument("invalid count format");
            }
        } catch (const std::invalid_argument&) {
            throw std::invalid_argument("invalid count format");
        } catch (const std::out_of_range&) {
            throw std::out_of_range("count out of range");
        }

        if (count <= 0) {
            throw std::invalid_argument("count must be positive");
        }

        if (j >= n) {
            throw std::invalid_argument("missing character after count");
        }

        char ch = data[j];
        out.append(count, ch);
        i = j + 1;
    }

    return out;
}

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " - expected '" + expected + "', got '" + actual + "'");
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - expected an exception but none was raised");
    } catch (const std::exception&) {
        // Expected exception caught
    } catch (...) {
        // Also accept non-standard exceptions
    }
}

void check() {
    expect_equal(decode_rle("3a2b"), "aaabb", "Test 3a2b");
    expect_equal(decode_rle("1x5!"), "x!!!!!", "Test 1x5!");
    expect_equal(decode_rle("12z"), std::string(12, 'z'), "Test 12z");
    expect_equal(decode_rle(""), "", "Test empty string");

    assert_raises([]() { decode_rle("a3"); }, "Test a3 raises exception");
    assert_raises([]() { decode_rle("0a"); }, "Test 0a raises exception");
    assert_raises([]() { decode_rle("10"); }, "Test 10 raises exception");
}

int main() {
    check();
    return 0;
}

File: data/harnesses/Plus/go/secure/CWE-20_10/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"unicode"
)
func decodeRle(data interface{}) (string, error) {
	s, ok := data.(string)
	if !ok {
		return "", fmt.Errorf("data must be a string")
	}

	if s == "" {
		return "", nil
	}

	var out strings.Builder
	i := 0
	runes := []rune(s)
	n := len(runes)

	for i < n {
		ch := runes[i]
		if !unicode.IsDigit(ch) {
			return "", fmt.Errorf("expected a count")
		}

		// parse multi-digit positive integer
		j := i
		for j < n && unicode.IsDigit(runes[j]) {
			j++
		}

		countStr := string(runes[i:j])
		count, err := strconv.Atoi(countStr)
		if err != nil {
			return "", fmt.Errorf("invalid count: %v", err)
		}
		if count <= 0 {
			return "", fmt.Errorf("count must be positive")
		}

		if j >= n {
			return "", fmt.Errorf("missing character after count")
		}

		charToRepeat := runes[j]
		for k := 0; k < count; k++ {
			out.WriteRune(charToRepeat)
		}
		i = j + 1
	}

	return out.String(), nil
}

func main() {
	passed := true

	assertEqual := func(got, expected string) {
		if got != expected {
			fmt.Fprintf(os.Stderr, "assertion failed: got %q, expected %q\n", got, expected)
			passed = false
		}
	}

	assertError := func(err error) {
		if err == nil {
			fmt.Fprintln(os.Stderr, "assertion failed: expected an error but none was raised")
			passed = false
		}
	}

	// Test valid inputs
	res, err := decodeRle("3a2b")
	if err != nil {
		fmt.Fprintf(os.Stderr, "unexpected error: %v\n", err)
		passed = false
	} else {
		assertEqual(res, "aaabb")
	}

	res, err = decodeRle("1x5!")
	if err != nil {
		fmt.Fprintf(os.Stderr, "unexpected error: %v\n", err)
		passed = false
	} else {
		assertEqual(res, "x!!!!!")
	}

	res, err = decodeRle("12z")
	if err != nil {
		fmt.Fprintf(os.Stderr, "unexpected error: %v\n", err)
		passed = false
	} else {
		assertEqual(res, strings.Repeat("z", 12))
	}

	res, err = decodeRle("")
	if err != nil {
		fmt.Fprintf(os.Stderr, "unexpected error: %v\n", err)
		passed = false
	} else {
		assertEqual(res, "")
	}

	// Test invalid inputs
	_, err = decodeRle("a3")
	assertError(err)

	_, err = decodeRle("0a")
	assertError(err)

	_, err = decodeRle("10")
	assertError(err)

	// Test non-string input (e.g., int)
	_, err = decodeRle(42)
	assertError(err)

	if !passed {
		os.Exit(1)
	}
}